### DISCUSSION TITLE
Are keys ordered?

### DISCUSSION THREAD

**Discussion by fithisux** [guycipher/k4#26] [General]
It is not clear from the documentation.

**Comment by guycipher** [comment]
Yep.  On skiplist and sstables.

**Reply by fithisux** [reply]
In other words K4 is an instance of this?
https://en.wikipedia.org/wiki/Ordered_Key-Value_Store

**Reply by guycipher** [reply]
That is correct.  The iterators scan in order. On flush and compaction the sstables are created using an iterator thus its contents are sorted as well.  K4 will try to gather keys in order but keep in mind the contents of sstables can vary so the iteration in a whole after the skiplist may not be in order.   You can play around with this.  The skiplist (memtable) is however absolutely it's own OKVS.  K4 in a whole may not be entirely because each sstable is a sorted set of key value pairs.

**Reply by fithisux** [reply]
The above link though mentions that a lot of Higher level abstractions use the Ordering of the Key-Values. You definitely need it if you need to do  O(logN) scans. Maybe I'm wrong. My personal project is to port LevelGraph/Level library on top of K4. Do you see it as feasible given the properties of K4?
